### ISSUE TITLE
find gtk-2.0 wildcard command hangs for ~8 seconds

### ISSUE THREAD

**Issue by CosineP** [ibus/ibus#2070]
Please fill in the following items if you don't know the root cause.

Which distribution and version?: Debian Buster (Testing)

Which desktop environment and version?: Gnome + AwesomeWM

Which session type?: X11

Which application and version?: ibus

IBus version?: 1.5.19-1

Issue description:  
something is executing this command: `find /usr/lib -path /usr/lib/*/gtk-2.0/*/immodules/im-ibus.so` which hangs for about 8 seconds, blocking everything. this happens when i log in to the session and ibus starts

Steps to reproduce:
1. start ibus (im using Japanese keyboard)

**Comment by CosineP** [comment]
I don't, ibus runs it

**Comment by fujiwarat** [comment]
I don't think so.

**Comment by HebaruSan** [comment]
I'm seeing this, too (Ubuntu 19.04). My first login after a reboot hangs and fails and drops me back to the login prompt, and it works afterwards, presumably thanks to disk caching. During the freeze, that exact `find` command is visible in the process listings.

The same pattern is used in `/etc/X11/xinit/xinput.d/ibus`, but there it's in a shell for loop glob instead of a `find` command:

```bash
for IM_CONFIG_MARKER in /usr/lib/*/gtk-2.0/*/immodules/im-ibus.so \
                        /usr/lib/gtk-2.0/*/immodules/im-ibus.so ; do
    if [ -e $IM_CONFIG_MARKER ]; then
        IM_CONFIG_MARKER2=1
        break
    fi
done
```

This had me wondering whether bash might defer globs to `find` in some cases, but that seems pretty unlikely. If ibus doesn't do this itself, maybe Debian/Ubuntu replicated the above logic in another place...?

**Comment by HebaruSan** [comment]
A few places where code in the ibus repo constructs strings that might be passed to such a find command, depending on the values of variables:

https://github.com/ibus/ibus/blob/1c54fea645b2ec3a385d492e4ca30a70dbb87c3a/ibus.spec.in#L313

https://github.com/ibus/ibus/blob/1c54fea645b2ec3a385d492e4ca30a70dbb87c3a/configure.ac#L230

**Comment by HebaruSan** [comment]
My greps of the source for ibus and libgtk came up empty handed, but I think this from `/usr/share/im-config/inpurrc.common` (from the `im-config` package) looks promising:

```bash
# imconfig_find_files_with_prefixpath_glob
#    check if files exist with given array of strings (shell glob acceptable)
#    accept a prefix where "find" is executed on, with a list of strings of
#    glob-enabled absolute file path.
#    return a string of matched files, using absolute path, separated with
#    newline. The string is not ended with newline.
#
#    Example:
#        imconfig_find_files_with_prefixpath_glob /usr/lib \
#            '/usr/lib/*/gtk-2.0/*/im_fcitx.so' \
#            '/usr/lib/gtk-2.0/*/im_fcitx.so'
imconfig_find_files_with_prefixpath_glob () {
    _PREFIX_PATH="$1"
    shift
    _RESULT=""
    for _STRING_WITH_GLOB in "$@"; do
        _TMP_RESULT="$(find "${_PREFIX_PATH}" -path "${_STRING_WITH_GLOB}")"
        if [ "x${_TMP_RESULT}" != "x" ]; then
            _RESULT="$(printf '%s\n%s' "${_RESULT}" "${_TMP_RESULT}")"
        fi
    done
    printf '%s' "${_RESULT}"
}
```

- It's a `find` command (line 6)
- It uses the `-path` parameter in the same format
- It's related to GTK and IM

This is called from with parameters of `/usr/lib` and `/usr/lib/*/gtk-2.0/*/immodules/im-ibus.so` in `/usr/share/im-config/data/21_ibus.rc` (again package `im-config`):

```bash
GTK_IM_MODULE=xim
# use immodule only when available for both GTK 2.0 and 3.0
IM_CONFIG_MARKER2=0
IM_CONFIG_MARKER="$(imconfig_find_files_with_prefixpath_glob /usr/lib \
        '/usr/lib/*/gtk-2.0/*/immodules/im-ibus.so' \
        '/usr/lib/gtk-2.0/*/immodules/im-ibus.so')"
if [ "x${IM_CONFIG_MARKER}" != "x" ]; then
    IM_CONFIG_MARKER2=1
fi
```

So it looks like we should be bothering the maintainers of `im-config`.

**Comment by HebaruSan** [comment]
Aha, that team already has a report!

https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=925160

Closed as fixed with some changes last month to switch from `find` to globbing, so now it's a matter of waiting for the fix to reach our distros. Hopefully that link will help others who find this issue by searching for the command as I did.
